src/Compilers/Shared/BuildServerConnection.cs

```
@@ -595,4 +621,205 @@ public static string GetTempPath(string workingDir)
+                    return true;
+                }
+                catch (IOException)
+                {
+                    // Lock currently held by someone else
+                    Thread.Sleep(0);
+                }
+                catch (Exception)
+                {
+                    // Something else went wrong.
+                    return false;
+                }
+            } while (sw.ElapsedMilliseconds < timeoutMs);
+
+            return false;
+        }
+
+        public void Unlock ()
+        {
+            if (!IsLocked)
+                return;
+            Stream.Unlock(0, 0);
+            IsLocked = false;
+        }
+
+        public void Dispose ()
+        {
+            var wasLocked = IsLocked;
+            if (wasLocked)
+                Unlock();
+            Stream.Dispose();
+            // We do not delete the lock file here because there is no reliable way to perform a
+            //  'delete if no one has the file open' operation atomically on *nix. This is a leak.
+        }
+    }
+
+    internal sealed class ServerNamedMutex : IServerMutex
+    {
+        private readonly Mutex Mutex;
+
+        public bool IsDisposed { get; private set; }
+        public bool IsLocked { get; private set; }
+
+        public ServerNamedMutex (string mutexName, bool createNewLocked, out bool createdNew)
+        {
+            if (createNewLocked) {
+                Mutex = new Mutex(
+                    initiallyOwned: true,
+                    name: mutexName,
+                    createdNew: out createdNew
+                );
+                if (createdNew)
+                    IsLocked = true;
+            } else {
+                Mutex.TryOpenExisting(mutexName, out Mutex);
+                createdNew = false;
+            }
+
+            IsDisposed = (Mutex == null);
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsDisposed)
+                return false;
```

Should throw `InvalidOperationException` here as well.